### PR TITLE
various things that allow miri to crash less and execute more code

### PR DESCRIPTION
**Pull request by oli-obk**

### CODE REVIEW

File: src/interpreter/mod.rs [hunk 1]
```diff
@@ -1093,12 +1093,20 @@ fn report(tcx: TyCtxt, ecx: &EvalContext, e: EvalError) {
...
             }
         }
-        err.span_note(span, &format!("inside call to {}", Instance(def_id, substs)));
+        let inst = Instance(def_id, substs);
+        match ::std::panic::catch_unwind(|| {
+            format!("inside call to {}", inst)
+        }) {
+            Ok(msg) => err.span_note(span, &msg),
+            Err(_) => err.span_note(span, &format!("ppaux::parameterized failed: {:?}, {:?}", def_id, substs)),
+        };
```
Comment (eddyb) [inline src/interpreter/mod.rs]: Whoaaaa what?! This is a bug in the compiler, please don't work around it.

File: src/interpreter/terminator.rs [hunk 2]
```diff
@@ -514,23 +526,18 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
...
-                // let fn_ptr_ty = match method_ty.sty {
-                //     ty::TyFnDef(_, _, fty) => tcx.mk_ty(ty::TyFnPtr(fty)),
-                //     _ => unreachable!("expected fn item type, found {}",
-                //                       method_ty)
-                // };
-                // Callee::ptr(immediate_rvalue(llfn, fn_ptr_ty))
+            traits::VtableFnPointer(fn_ty) => {
+                if let ty::TyFnDef(did, ref substs, _) = fn_ty.fn_ty.sty {
+                    args.remove(0);
+                    Ok((did, substs))
```
Comment (eddyb) [inline src/interpreter/terminator.rs]: Heh, nice.

File: src/interpreter/mod.rs (hunk 1 above)
Comment (solson) [inline src/interpreter/mod.rs]: I've been hitting this bug, too, but I haven't investigated it yet.

File: src/interpreter/terminator.rs [hunk 3]
```diff
@@ -514,23 +526,18 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
...
-                // let llfn = trans_fn_pointer_shim(ccx, trait_closure_kind, fn_ty);
-
-                // let method_ty = def_ty(tcx, def_id, substs);
-                // let fn_ptr_ty = match method_ty.sty {
-                //     ty::TyFnDef(_, _, fty) => tcx.mk_ty(ty::TyFnPtr(fty)),
-                //     _ => unreachable!("expected fn item type, found {}",
-                //                       method_ty)
-                // };
-                // Callee::ptr(immediate_rvalue(llfn, fn_ptr_ty))
+            traits::VtableFnPointer(fn_ty) => {
```
Comment (solson) [inline src/interpreter/terminator.rs]: Rename this `fn_ty` to `vtable_fn_ptr`. (It has type [`VtableFnPointerData`](http://manishearth.github.io/rust-internals-docs/rustc/traits/struct.VtableFnPointerData.html).)

File: src/interpreter/terminator.rs (hunk 2 above)
Comment (solson) [inline src/interpreter/terminator.rs]: :+1: Passing in `&mut args` is what I planned to do for this as well.

File: src/interpreter/mod.rs (hunk 1 above)
Comment (oli-obk) [inline src/interpreter/mod.rs]: I'm fixing this bug independently. It was a blocker, and I didn't want to wait the 2 hours for a debug stage1. I have the debug stage1 now, and will investigate.

Comment (solson) [inline src/interpreter/mod.rs]: I will merge this PR as-is.

**Pull request merged**